keps/sig-node/4815-dra-partitionable-devices/README.md

```diff
@@ -0,0 +1,2140 @@
+scheduler. However, as soon as its capacity has been pulled down by any given
+"sink" device, it can no longer be allocated until its capacity is freed again.
+
+Likewise, so long as all of the advertised capacity of a "sink" device can be
+satisfied by the set of "source" devices it references in its
+`ConsumesCapacityFrom` field, it is free to be allocated by the scheduler.
+However, if any of its advertised capacity cannot be satisfied by one of its
+referenced "source" devices, then it cannot be allocated until that capacity is
+freed by some other device.
+
+Note that in order to support nested partitioning, a "sink" device *may*
+provide a reference to another "sink" device in its `ConsumesCapacityFrom`
+field, so long as:
+
+1. Each device along the recursive chain of references is able to pull enough
+   capacity from the devices in its own `ConsumesCapacityFrom` list to satisfy
+   its allocation.
+
+1. The final device in the chain is a "source" device.
+
+When such a device is allocated, the scheduler will need to track the full
+capacity required to satisfy each of the sink devices along the chain. In this
+way, all intermediate sink devices will essentially be rendered
+"unschedulable", with the last-level sink device pulling its capacity from the
+devices it references directly.
+
+### Putting it all together for the MIG use-case
+
+The example in the previous section only lists a **single** "memory" capacity
+on "gpu-0" that gets divided evenly across 4 possible sub-partitions. While
+this simple use case is certainly supported, it is not representative of the
+more complex use-cases we envision the extensions in this KEP to be used for.
+Specifically, for partitioning devices across multiple dimensions, such as is
+necessary to support the MIG use-case for NVIDIA GPUs described previously.
+
+With MIG, one can think of a GPU as a "bag" of resources that can be pulled
+together in a discrete set of ways to construct a MIG device with a particular
+"profile". These resources include things like "memory", "number of JPEG
+engines", "discrete memory slices on the physical GPU die", etc.
+
+Using the idea of "sink" and "source" devices, we can construct a full GPU as a
+"source" device with a set of capacities representing the different resources
+in its "bag" of resources. Each MIG device then becomes a "sink" device listing
+out the capacities it requires to build out its specific "profile" and then
+referencing the full GPU it is attached to in its `ConsumesCapacityFrom` field.
+
+For example, an unrolled "source" device for an NVIDIA A100 GPU looks as follows:
+```yaml
+- name: gpu-0
+  composite:
+    attributes:
+      ...
+    capacity:
+      copy-engines:
+        quantity: "7"
+      decoders:
+        quantity: "5"
+      encoders:
+        quantity: "0"
+      jpeg-engines:
+        quantity: "1"
+      memory:
+        quantity: 40Gi
+      memorySlice0:
```

Comment:
Agreed -- but this is where we are at the moment with the design.

I had also talked about introducing an `intRange` abstraction at one point so that one could directly write:
```
consumes:
  memorySlices: "0-3"
```
but that had it's own problems.